pkg/controllers/br/tibrgc/tasks/ctx.go

```diff
@@ -0,0 +1,246 @@
+	"github.com/pingcap/tidb-operator/pkg/client"
+	t "github.com/pingcap/tidb-operator/pkg/utils/task/v3"
+)
+
+type ReconcileContext struct {
+	namespacedName types.NamespacedName
+	ctx            context.Context
+	cli            client.Client
+
+	tibrgc    *v1alpha1br.TiBRGC
+	cluster   *v1alpha1.Cluster
+	t2Cronjob *batchv1.CronJob
+	t3Cronjob *batchv1.CronJob
+
+	// tibrgcList4SameCluster is a list of tibrgc instances in the same cluster
+	// this filed only used to prevent another tibrgc instance scheduled in the same cluster
+	tibrgcList4SameCluster *v1alpha1br.TiBRGCList
+}
+
+func NewReconcileContext(ctx context.Context,
+	namespacedName types.NamespacedName,
+	cli client.Client,
+) *ReconcileContext {
+	return &ReconcileContext{
+		namespacedName: namespacedName,
+		ctx:            ctx,
+		cli:            cli,
+	}
+}
+
+func (c *ReconcileContext) NamespacedName() types.NamespacedName {
+	return c.namespacedName
+}
+
+func (c *ReconcileContext) Client() client.Client {
+	return c.cli
+}
+
+func (c *ReconcileContext) TiBRGC() *v1alpha1br.TiBRGC {
+	return c.tibrgc
+}
+
+func (c *ReconcileContext) TiBRGCList4SameCluster() *v1alpha1br.TiBRGCList {
+	return c.tibrgcList4SameCluster
+}
+
+func (c *ReconcileContext) T2Cronjob() *batchv1.CronJob {
+	return c.t2Cronjob
+}
+
+func (c *ReconcileContext) T3Cronjob() *batchv1.CronJob {
+	return c.t3Cronjob
+}
+
+func (c *ReconcileContext) Cluster() *v1alpha1.Cluster {
+	return c.cluster
+}
+
+func (c *ReconcileContext) TLSEnabled() bool {
+	return c.Cluster().Spec.TLSCluster != nil && c.Cluster().Spec.TLSCluster.Enabled
+}
+
+func (c *ReconcileContext) RefreshTiBRGC() error {
+	tibrgc := &v1alpha1br.TiBRGC{}
+	if err := c.cli.Get(c.ctx, c.namespacedName, tibrgc); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+		c.tibrgc = nil
+		return nil // Not found is not an error
+	}
+	c.tibrgc = tibrgc
+	return nil
+}
+
+func (c *ReconcileContext) RefreshTiBRGCList4SameCluster() error {
+	if c.tibrgc == nil {
+		return errors.New("TiBRGC is nil")
+	}
+
+	var brgclist v1alpha1br.TiBRGCList
+	if err := c.cli.List(c.ctx, &brgclist, &client.ListOptions{
+		Namespace: c.tibrgc.Namespace,
+		FieldSelector: fields.SelectorFromSet(map[string]string{
+			ClusterNameField: c.tibrgc.Spec.Cluster.Name,
+		}),
+	}); err != nil {
+		return err
+	}
+	c.tibrgcList4SameCluster = &brgclist
+	return nil
+}
+
+func (c *ReconcileContext) RefreshCluster() error {
+	if c.tibrgc == nil {
+		return errors.New("TiBRGC is nil")
+	}
+	clusterName := types.NamespacedName{
+		Namespace: c.tibrgc.Namespace,
+		Name:      c.tibrgc.Spec.Cluster.Name,
+	}
+	cluster := &v1alpha1.Cluster{}
+	if err := c.cli.Get(c.ctx, clusterName, cluster); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return err
+		}
+		c.cluster = nil
+		return nil // Not found is not an error
+	}
+	c.cluster = cluster
+	return nil
+}
+
+func (c *ReconcileContext) RefreshT2CronJob() error {
```

Comment:
Because the controller owns CronJob, if the CronJob is created/updated/deleted, a reconcile of TiBRGC is triggered.